rewrite-java/src/main/java/org/openrewrite/java/tree/MethodTypeBuilder.java

```diff
@@ -15,14 +15,34 @@
  */
 package org.openrewrite.java.tree;
 
-import com.koloboke.collect.set.hash.HashObjSet;
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import org.openrewrite.Incubating;
+import org.openrewrite.internal.StringUtils;
 
 import java.util.*;
 import java.util.stream.Collectors;
 
+/**
+ * This builder provides a fluent API for constructing method types. This builder does not support separate generic
+ * and resolved methods signatures.
+ * <P>
+ * <PRE>
+ *     EXAMPLE:
+ *
+ *     import static org.openrewrite.java.tree.MethodTypeBuilder.methodType;
+ *     .
+ *     .
+ *     .
+ *     JavaType.Method method = methodType()
+ *             .declaringClass("org.assertj.core.api.Assertions")
+ *             .flags(Flag.Public, Flag.Static)
+ *             .returnType("org.assertj.core.api.AbstractBooleanAssert")
+ *             .name("assertThat")
+ *             .parameter("boolean", "actual")
+ *             .build();
```

Comment:
Can you compose these builders? Say, to create a chain of method invocations like `foo().bar().baz()`?

I vaguely remember that a chain of such method calls was part of what sent you down this path in the first place but I don't see  `select(MethodBuilder)` or `select(J.MethodInvocation)` methods on this class.

Reply:
yep, that is actually a TODO: Add a MethodInvocationBuilder and then you can chain these things together: 

```
import static org.openrewrite.java.tree.MethodInvocationBuilder.newMethodInvocation;
import static org.openrewrite.java.tree.MethodTypeBuilder.newMethodType;
.
.
.
  J.MethodInvocation invocation = newMethodInvocation()
    .randomId()
    .select(...)
    .methodType(newMethodType()
      .declaringClass("org.assertj.core.api.Assertions")
      .flags(Flag.Public, Flag.Static)
      .returnType("org.assertj.core.api.AbstractBooleanAssert")
      .name("assertThat")
      .parameter("boolean", "actual")
      .build()
    )
   .bulid();
```

Reply:
Cool cool. One thing to consider in creating this builder is: which direction should it go?
A programmer writes `foo().bar().baz()`, but the AST goes `baz( select: bar( select: foo()))`. Should this builder correspond more closely to a programmer's intuition or to the underlying data structure?